vdr/didweb/web.go

```diff
@@ -0,0 +1,116 @@
+package didweb
+
+import (
+	"crypto/tls"
+	"errors"
+	"fmt"
+	"github.com/nuts-foundation/go-did/did"
+	"github.com/nuts-foundation/nuts-node/vdr/types"
+	"io"
+	"mime"
+	"net/http"
+	"net/url"
+	"strings"
+	"time"
+)
+
+// MethodName is the DID method name used by did:web
+const MethodName = "web"
+
+var _ types.DIDResolver = (*Resolver)(nil)
+
+// Resolver is a DID resolver for the did:web method.
+type Resolver struct {
+	HttpClient *http.Client
+}
+
+// NewResolver creates a new did:web Resolver with default TLS configuration.
+func NewResolver() *Resolver {
+	transport := http.DefaultTransport
+	if httpTransport, ok := transport.(*http.Transport); ok {
+		// Might not be http.Transport in testing
+		httpTransport = httpTransport.Clone()
+		httpTransport.TLSClientConfig = &tls.Config{
+			MinVersion: tls.VersionTLS12,
+		}
+		transport = httpTransport
+	}
+	return &Resolver{
+		HttpClient: &http.Client{
+			Transport: transport,
+			Timeout:   5 * time.Second,
+		},
+	}
+}
+
+// Resolve implements the DIDResolver interface.
+func (w Resolver) Resolve(id did.DID, _ *types.ResolveMetadata) (*did.Document, *types.DocumentMetadata, error) {
+	if id.Method != "web" {
+		return nil, nil, errors.New("DID is not did:web")
+	}
+
+	var baseID = id.ID
+	var path string
+	subpathIdx := strings.Index(id.ID, ":")
+	if subpathIdx == -1 {
+		path = "/.well-known/did.json"
+	} else {
+		// subpaths are encoded as / -> :
+		baseID = id.ID[:subpathIdx]
+		path = id.ID[subpathIdx:]
+		path = strings.ReplaceAll(path, ":", "/") + "/did.json"
+	}
+
+	unescapedID, err := url.PathUnescape(baseID)
+	if err != nil {
+		return nil, nil, fmt.Errorf("invalid did:web: %w", err)
+	}
+	targetURL := "https://" + unescapedID + path
+
+	// TODO: Support DNS over HTTPS (DOH), https://www.rfc-editor.org/rfc/rfc8484
+	httpResponse, err := w.HttpClient.Get(targetURL)
+	if err != nil {
+		return nil, nil, fmt.Errorf("did:web HTTP error: %w", err)
+	}
+	defer httpResponse.Body.Close()
+	if !(httpResponse.StatusCode >= 200 && httpResponse.StatusCode < 300) {
+		return nil, nil, fmt.Errorf("did:web non-ok HTTP status: %s", httpResponse.Status)
+	}
+
+	ct, _, err := mime.ParseMediaType(httpResponse.Header.Get("Content-Type"))
+	if err != nil {
+		return nil, nil, fmt.Errorf("did:web invalid content-type: %w", err)
+	}
+	switch ct {
```

Comment:
§2.4 of the [spec](https://w3c-ccg.github.io/did-method-web/#key-material-and-document-handling) expects most `/did.json` to be served as `application/json` (it's in the file extension), including LD documents. So LD documents are detected based on the presence of an `@context`. Can we process it using these rules?

Reply:
We _could_ process them (and it would probably make the Nuts Node a purer did:web resolver), but we don't need it for our use cases (see comment on line 86). And we don't do it for `did:nuts` documents anyways, so at least the behavior is the same for how we treat `did:web`.

Reply:
The spec also mentions rejecting any document containing `@context` that fail JSON-LD parsing. Ignoring this only means we are a bit relaxed on validation? Probably fine as long as we don't serve did:web documents later on that adopt this relaxed view.